File: selenium/src/csharp/WebdriverBackedSelenium/Internal/SeleniumEmulation/KeyEvent.cs
using System;
using System.Collections.Generic;
using System.Text;
using OpenQA.Selenium;

namespace Selenium.Internal.SeleniumEmulation
{
    class KeyEvent : SeleneseCommand
    {
        private ElementFinder finder;
        private JavaScriptLibrary library;
        private KeyState keyState;
        private string eventName;

        public KeyEvent(ElementFinder elementFinder, JavaScriptLibrary js, KeyState state, String eventName)
        {
            this.finder = elementFinder;
            this.library = js;
            this.keyState = state;
            this.eventName = eventName;
        }

        protected override object HandleSeleneseCommand(IWebDriver driver, string locator, string value)
        {
            object[] parameters = new object[] 
            { 
                value, 
                keyState.ControlKeyDown, 
                keyState.AltKeyDown, 
                keyState.ShiftKeyDown, 
                keyState.MetaKeyDown 
            };
            library.CallEmbeddedSelenium(driver, eventName, finder.FindElement(driver, locator), parameters);

            return null;
        }
    }
}

File: java/client/src/com/thoughtworks/selenium/webdriven/commands/KeyEvent.java
// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package com.thoughtworks.selenium.webdriven.commands;

import com.thoughtworks.selenium.webdriven.ElementFinder;
import com.thoughtworks.selenium.webdriven.JavascriptLibrary;
import com.thoughtworks.selenium.webdriven.SeleneseCommand;

import org.openqa.selenium.WebDriver;

public class KeyEvent extends SeleneseCommand<Void> {
  private final ElementFinder elementFinder;
  private final JavascriptLibrary js;
  private final KeyState state;
  private final String eventName;

  public KeyEvent(ElementFinder elementFinder, JavascriptLibrary js, KeyState state,
      String eventName) {
    this.elementFinder = elementFinder;
    this.js = js;
    this.state = state;
    this.eventName = eventName;
  }

  @Override
  protected Void handleSeleneseCommand(WebDriver driver, String locator, String value) {
    js.callEmbeddedSelenium(driver, eventName, elementFinder.findElement(driver, locator),
        value, state.controlKeyDown, state.altKeyDown, state.shiftKeyDown, state.metaKeyDown);

    return null;
  }
}
